services/shell/copy_local.go

```diff
@@ -0,0 +1,443 @@
+			if parentInfo.IsDir() {
+				// it is a directory, then put the source/file directory in it
+				// destination becomes parent
+				rename = true
+			} else {
+				// it is a file, then error
+				return nil, fmt.Errorf("parent of destination %q is an existing file, not a directory", f.destination)
+			}
+		}
+
+		if rename {
+			overrideName = filepath.Base(f.destination)
+			finalDestination = filepath.Dir(f.destination)
+		}
+	case CopyFilesSourceTypeMultipleUnknown:
+		fallthrough
+	default:
+		return nil, fmt.Errorf("do not know how to process source copy type %q", sourceType)
+	}
+	return &localFileCopier{
+		sourceType:   sourceType,
+		dst:          finalDestination,
+		overrideName: overrideName,
+		preserve:     f.preserve,
+	}, nil
+}
+
+func (f *localFileCopyFactory) Close(ctx context.Context) error {
+	return nil
+}
+
+// A localFileCopier takes in files and copies them to a set destination. It should be created
+// with a localFileCopyFactory.
+type localFileCopier struct {
+	sourceType   CopyFilesSourceType
+	dst          string
+	overrideName string
+	preserve     bool
+}
+
+func (copier *localFileCopier) Copy(ctx context.Context, file File) error {
+	defer func() {
+		utils.UncheckedError(file.Data.Close())
+	}()
+
+	fileName := file.RelativeName
+	if copier.overrideName != "" {
+		// only change the first part of the file name for directory
+		// renaming purposes. This is based on SCP logic.
+		fileSplit := splitPath(file.RelativeName)
+		if len(fileSplit) == 0 {
+			fileSplit = []string{copier.overrideName}
+		} else {
+			fileSplit[0] = copier.overrideName
+		}
+		fileName = filepath.Join(fileSplit...)
+	}
+	fullPath := filepath.Join(copier.dst, fileName)
+
+	fileInfo, err := file.Data.Stat()
+	if err != nil {
+		return err
+	}
+	if fileInfo == nil {
+		return fmt.Errorf("expected file info for %q to be non-nil", fileName)
+	}
+
+	parentPath := filepath.Dir(fullPath)
+	if parentPath == "" {
+		return fmt.Errorf("expected non-empty parent path to destination %q", copier.dst)
+	}
+	if fileInfo, err := os.Stat(parentPath); err != nil {
+		if !errors.Is(err, fs.ErrNotExist) {
+			return err
+		}
+		// this will later be updated with chmod for a specific directory. It's safe to make
+		// directories here because we assume whoever constructed us has validated the top-level
+		// directory as existing or created.
+		//nolint:gosec // this is from an authenticated/authorized connection
+		if err := os.MkdirAll(parentPath, 0o755); err != nil {
+			return err
+		}
+	} else if fileInfo == nil {
+		return fmt.Errorf("expected file info for %q to be non-nil", parentPath)
+	} else if !fileInfo.IsDir() {
+		return fmt.Errorf("invariant: parent path %q should have been a directory", parentPath)
+	}
+
+	var fileMode fs.FileMode
+	modTime := time.Now()
+	switch {
+	case copier.preserve:
+		modTime = fileInfo.ModTime()
+		fileMode = fileInfo.Mode()
+	case fileInfo.IsDir():
+		fileMode = 0o755
+	default:
+		fileMode = 0o644
```

Comment:
[q] What are these permission here and above? And, why do they differ depending on directory vs file?

Reply:
Above refers to the preserve flag in the proto (preserve indicates the the receiver should provide the metadata in the file to reflect the same state in the sender's filesystem as applicable.). Below are the defaults modes for directories (rwx, rx, rx) and regular files (rw, r, r). going to change the global side (last part of tuple) to 0 though actually. FYI: the x/executable bit on a directory is not so you can execute a directory but so that you can list files (I don't make the rules :( )

Reply:
Cool thanks for context.

> the x/executable bit on a directory is not so you can execute a directory but so that you can list files 

TIL 😆 . Just tried it out locally `ls: test: Permission denied`.